File: src/python/MatrixMult1.py
#
# MatrixMult1.py
#
# Created by Jukka Aro on 13.3.2012.
#

import sys

def read_input(input_file, dim, is_float):
	f = open(input_file, 'r')
	M = [0]*dim*dim
	if is_float:
		for i in range (0,dim):
			line = f.readline()
			tokens  = line.split()
			for j in range(0,dim):
				M[i*dim+j] = float(tokens[j])
	else:
		for i in range (0,dim):
			line = f.readline()
			tokens  = line.split()
			for j in range(0,dim):
				M[i*dim+j] = int(float(tokens[j]))
	f.close()
	return M
	
def multiply_matrix(In, dim, is_float):
	Out = [0]*dim*dim
	
	for i in range (0,dim):
		for j in range (0,dim):
			res = 0
			for k in range (0,dim):
				res += In[i*dim+k]*In[k*dim+j]
			Out[i*dim+j] = res

#program entry:

if len(sys.argv) < 3 or len(sys.argv) > 5:
	print "Illegal numger of arguments"
	sys.exit()

input_file = sys.argv[1]
dim = int(sys.argv[2])
is_float = False
dry_run = False

if len(sys.argv) == 4:
	if sys.argv[3] == '--float':
		is_float = True
	elif sys.argv[3] == '--dry-run':
		dry_run = True
	else:
		print "Illegal argument: " + sys.argv[3]
		sys.exit()
if len(sys.argv) == 5:
	if sys.argv[3] == '--float' and sys.argv[4] == '--dry-run':
		is_float = True
		dry_run = True
	else:
		print "Illegal arguments: " + sys.argv[3] + " " + sys.argv[4]
		sys.exit()

if dry_run:
	read_input(input_file, dim, is_float)
else:
	multiply_matrix(read_input(input_file, dim, is_float), dim, is_float)

File: src/perl/MatrixMult1.pl
#
# MatrixMult1.pl
#
# Created by Jukka Aro on 14.3.2012.
#

sub read_input{
	open FILE, $_[0] or die $!;
	@in_matrix = 0 x ($dim*$dim);
	@lines = <FILE>;

	if($_[1] == 1){
		for($i = 0; $i < $dim; $i++) {
			@tokens = split(' ', $lines[$i]);
			for($j = 0; $j < $dim; $j++) {
				$in_matrix[$dim*$i+$j] = $tokens[$j];
			}
		}
	}
	else{
		for($i = 0; $i < $dim; $i++) {
			@tokens = split(' ', $lines[$i]);
			for($j = 0; $j < $dim; $j++) {
				$in_matrix[$dim*$i+$j] = int($tokens[$j]);
			}
		}
	}
	
	close(FILE);
	@in_matrix;
}

sub multiply_matrix{
	@out_matrix = 0 x ($dim*$dim);

	for($i = 0; $i < $dim; $i++) {
		for($j = 0; $j < $dim; $j++) {
			$res = 0;
			for($k = 0; $k < $dim; $k++) {
				$res += $_[$dim*$i+$k]*$_[$dim*$k+$j];
			}
			$out_matrix[$dim*$i+$j] = $res;
		}
	}
}


#program entry:

if($#ARGV < 1 || $#ARGV > 3){
	die "Illegal number of arguments";
}

$is_float = 0;
$dry_run = 0;

if($#ARGV == 2){
	if($ARGV[2] eq "--float"){
		$is_float = 1;
	}
	elsif($ARGV[2] eq "--dry-run"){
		$dry_run = 1;
	}
	else{
		die "Illegal argument: ".$ARGV[2];
	}
}
if($#ARGV == 3){
	if($ARGV[2] eq "--float" && $ARGV[3] eq "--dry-run"){
		$is_float = 1;
		$dry_run = 1;
	}
	else{
		die "Illegal arguments: ".$ARGV[2]." ".$ARGV[3];
	}
}

$dim = $ARGV[1];

if($dry_run == 1){
	read_input($ARGV[0], $is_float);
}
else{
	multiply_matrix(read_input($ARGV[0], $is_float));
}